src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Resx = Microsoft.NetCore.Analyzers.MicrosoftNetCoreAnalyzersResources;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class SealInternalTypes : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1852";
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            Resx.CreateLocalizableResourceString(nameof(Resx.SealInternalTypesTitle)),
+            Resx.CreateLocalizableResourceString(nameof(Resx.SealInternalTypesMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeHidden_BulkConfigurable,
+            Resx.CreateLocalizableResourceString(nameof(Resx.SealInternalTypesDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false,
+            isReportedAtCompilationEnd: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            INamedTypeSymbol? comImportAttributeType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesComImportAttribute);
+
+            var candidateTypes = PooledConcurrentSet<INamedTypeSymbol>.GetInstance(SymbolEqualityComparer.Default);
+            var baseTypes = PooledConcurrentSet<INamedTypeSymbol>.GetInstance(SymbolEqualityComparer.Default);
+
+            context.RegisterSymbolAction(context =>
+            {
+                var type = (INamedTypeSymbol)context.Symbol;
+
+                if (type.TypeKind is TypeKind.Class &&
+                    !type.IsAbstract &&
+                    !type.IsStatic &&
+                    !type.IsSealed &&
+                    !type.IsExternallyVisible() &&
+                    (comImportAttributeType is null || !type.HasAttribute(comImportAttributeType)))
```

Comment:
HasAttribute already does null check:

https://github.com/dotnet/roslyn-analyzers/blob/334bb06a340f279d0c6de7d2db54a5cefb0506e0/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L642-L645

```suggestion
                    !type.HasAttribute(comImportAttributeType))
```